Rewrite fit_simulation.md for clarity. (#197)
SciML/SciMLDocs: docs/src/getting_started/fit_simulation.md

we will calculate the sum of squared errors as our loss metric.
The sum of squares can be quickly written in Julia via `sum(abs2,x)`.
Using this information, our loss function looks like:

```@example odefit
function loss(newp)
    newprob = remake(prob, p = newp)
    sol = solve(newprob, saveat = 1)
    loss = sum(abs2, sol .- xy_data)
    return loss, sol
end
```

Notice that our loss function returns the loss value as the first return,
but returns extra information (the ODE solution with the new parameters)
as an extra return argument.
We will explain why this extra return information is helpful in the next section.

### Step 5: Solve the Optimization Problem

This step will look very similar to [the first optimization tutorial](@ref first_opt),
except now we have a new loss function `loss` which returns both the loss value
and the associated ODE solution.
(In the previous tutorial, `L` only returned the loss value.)
The `Optimization.solve` function can accept an optional callback function
to monitor the optimization process using extra arguments returned from `loss`.

The callback syntax is always:

```
callback(

>     optimization function parameters,

Reviewer: ```suggestion
    optimization variables,
```

>     value being optimized,

Reviewer: ```suggestion
    the current loss value,
```

    other arguments returned from the loss function, ...
)
```

In this case, we will provide the callback the arguments `(_, l, sol)`,

> since there are no additional optimization function parameters.

Reviewer: This is incorrect since they are the optimization variables as changed above

Author: The "variable" wording seems too vague. There are "state variables", "parameters", and "hyperparameters" which are all variables in some sense. There are two different equation contexts where those definitions change, as I tried to explain at the bottom, and that creates a lot of the confusion for this passage.

I don't really understand what is incorrect about this. `p = [\alpha, \beta, \gamma, \delta]` are not the optimization parameters in this context. There are no optimization parameters which is why passing an underscore works.

Reviewer 2: We should probably make a table.

Typically the `return` statement would monitor the loss value
and stop once some criteria is reached, e.g. `return loss < 0.0001`,
but we will stop after a set number of iterations instead.
More details about callbacks in Optimization.jl can be found
[here](https://docs.sciml.ai/Optimization/stable/API/solve/).

```@example odefit
function callback(_, l, sol)
    display(l)
    plt = plot(sol, ylim = (0, 6), label = ["Current x Prediction" "Current y Prediction"])
    scatter!(plt, t_data, xy_data', label = ["x Data" "y Data"])
    display(plt)
    return false
end
```

With this callback function, every step of the optimization will display both the loss value and a plot of how the solution compares to the training data.

Now, just like [the first optimization tutorial](@ref first_opt),
we set up our `OptimizationFunction` and `OptimizationProblem`,
and then `solve` the `OptimizationProblem`.
We will initialize the `OptimizationProblem` with the same `pguess` we used
when setting up the ODE Model in Step 3.
Observe how `Optimization.solve` brings the model closer to the experimental data as it iterates towards better ODE parameter values!

> 

Reviewer: ```suggestion

Note that we are using the `PolyOpt()` solver choice here which is discussed https://docs.sciml.ai/Optimization/dev/optimization_packages/polyopt/ since parameter estimation of non-linear differential equations is generally a non-convex problem so we want to run a stochastic algorithm (Adam) to get close to the minimum and then finish off with a quasi-newton method (L-BFGS) to find the optima.
```

tspan = (0.0, 10.0)

# LV equation parameter. p = [α, β, δ, γ]
pguess = [1.0, 1.2, 2.5, 1.2]
```

Now we bring these pieces all together to define the `ODEProblem` and solve it.
Note that we solve this equation with the keyword argument `saveat = 1`
so that it saves a point at every ``\Delta t = 1`` to match our experimental data.

```@example odefit
# Set up the ODE problem with our guessed parameter values
prob = ODEProblem(lotka_volterra!, u0, tspan, pguess)

# Solve the ODE problem with our guessed parameter values
initial_sol = solve(prob, saveat = 1)

# View the guessed model solution
plt = plot(initial_sol, label = ["x Prediction" "y Prediction"])
scatter!(plt, t_data, xy_data', label = ["x Data" "y Data"])
```

> Clearly the parameter values that we guessed are not correct to model this system.
> However, we can use Optimization.jl together with DifferentialEquations.jl
> to fit our parameters to our training data.

Reviewer 2: ```suggestion
    Clearly the parameter values that we guessed are not correct to model this system.
    However, we can use Optimization.jl together with DifferentialEquations.jl
    to fit our parameters to our training data.
```

there is a useful functionality in the
[SciML problems interface](https://docs.sciml.ai/DiffEqDocs/stable/basics/problem/#Modification-of-problem-types)
called `remake` which creates a new version of an existing `SciMLProblem` with the aspect you want changed.
For example, if we wanted to change the initial condition `u0` of our ODE, we could do `remake(prob, u0 = newu0)`
For our case, we want to change around just the parameters, so we can do `remake(prob, p = newp)`. It is faster to `remake` an existing `SciMLProblem` than to create a new problem every iteration.

!!! note

    `remake` can change multiple items at once by passing more keyword arguments, i.e., `remake(prob, u0 = newu0, p = newp)`.
    This can be used to extend the example to simultaneously learn the initial conditions and parameters!

Now use `remake` to build the loss function. After we solve the new problem,
we will calculate the sum of squared errors as our loss metric.
The sum of squares can be quickly written in Julia via `sum(abs2,x)`.
Using this information, our loss function looks like:

```@example odefit
function loss(newp)
    newprob = remake(prob, p = newp)
    sol = solve(newprob, saveat = 1)

>     loss = sum(abs2, sol .- xy_data)
>     return loss, sol

Reviewer 2: ```suggestion
    l = sum(abs2, sol .- xy_data)
    return l, sol
```

end
```

Notice that our loss function returns the loss value as the first return,
but returns extra information (the ODE solution with the new parameters)
as an extra return argument.
We will explain why this extra return information is helpful in the next section.

### Step 5: Solve the Optimization Problem

This step will look very similar to [the first optimization tutorial](@ref first_opt),
except now we have a new loss function `loss` which returns both the loss value
and the associated ODE solution.
(In the previous tutorial, `L` only returned the loss value.)
The `Optimization.solve` function can accept an optional callback function
to monitor the optimization process using extra arguments returned from `loss`.

The callback syntax is always:

```
callback(
    optimization variables,
    the current loss value,
    other arguments returned from the loss function, ...
)
```

> In this case, we will provide the callback the arguments `(_, l, sol)`,
> since there are no additional optimization function parameters.

Reviewer 2: ```suggestion
In this case, we will provide the callback the arguments `(p, l, sol)`,
since it always takes the current state of the optimization first (`p`)
then the returns from the loss function (`l, sol`).
```

except now we have a new loss function `loss` which returns both the loss value
and the associated ODE solution.
(In the previous tutorial, `L` only returned the loss value.)
The `Optimization.solve` function can accept an optional callback function
to monitor the optimization process using extra arguments returned from `loss`.

The callback syntax is always:

```
callback(
    optimization variables,
    the current loss value,
    other arguments returned from the loss function, ...
)
```

In this case, we will provide the callback the arguments `(p, l, sol)`,
since it always takes the current state of the optimization first (`p`)
then the returns from the loss function (`l, sol`).
The return value of the callback function should default to `false`.
`Optimization.solve` will halt if/when the callback function returns `true` instead.
Typically the `return` statement would monitor the loss value
and stop once some criteria is reached, e.g. `return loss < 0.0001`,
but we will stop after a set number of iterations instead.
More details about callbacks in Optimization.jl can be found
[here](https://docs.sciml.ai/Optimization/stable/API/solve/).

```@example odefit

> function callback(_, l, sol)

Reviewer 2: ```suggestion
function callback(p, l, sol)
```

    DifferentialEquations.jl:

    ```math
    \frac{du}{dt} = f(u,p,t)
    ```

    - `f` in `ODEProblem` is the function defining the derivative `du` in the ODE.

        Here: `lotka_volterra!`

    - `u` in `ODEProblem` contains the state variables of `f`.

        Here: `x` and `y`

    - `p` in `ODEProblem` contains the parameter variables of `f`.

        Here: `\alpha`, `\beta`, `\gamma`, and `\delta`

    - `t` is the independent (time) variable.

        Here: indirectly defined with `tspan` in `ODEProblem` and `saveat` in `solve`

    Optimization.jl:

    ```math
    \min_{u} f(u,p)
    ```

    - `f` in `OptimizationProblem` is the function to minimize (optimize).

        Here: the [anonymous function](https://docs.julialang.org/en/v1/manual/functions/#man-anonymous-functions) `(x, _) -> loss(x)`

    - `u` in `OptimizationProblem` contains the state variables of `f` to be optimized.

        Here: the ODE parameters `\alpha`, `\beta`, `\gamma`, and `\delta` stored in `p`

    - `p` in `OptimizationProblem` contains any fixed
    [hyperparameters](https://en.wikipedia.org/wiki/Hyperparameter_(machine_learning)) of `f`.

>         Here: our `loss` function does not require any hyperparameters, so we pass `_` for this `p`.

Author: This line needs changed now to be consistent; `_` was changed to `p` above.

I used `_` and wrote this section in an attempt to avoid the confusion I had between this `p` and  `p = [\alpha, \beta, \gamma, \delta]` which is actually passed as `u`.

Reviewer 2: And this just underwent a change anyways. @Vaibhavdixit02 can you update this as part of that whole update?

Author: Should instead say something like " Here: our `loss` function does not require any hyperparameters, so we do not provide the `p` argument to `OptimizationProblem`."
